test/core/event_engine/work_queue_test.cc

```diff
@@ -0,0 +1,176 @@
+  ASSERT_TRUE(popped.has_value());
+  (*popped)->Run();
+  ASSERT_TRUE(ran);
+  ASSERT_TRUE(queue.Empty());
+}
+
+TEST(WorkQueueTest, BecomesEmptyOnPopBack) {
+  WorkQueue queue;
+  bool ran = false;
+  queue.Add([&ran] { ran = true; });
+  ASSERT_EQ(queue.Size(), 1);
+  ASSERT_FALSE(queue.Empty());
+  absl::optional<EventEngine::Closure*> closure = queue.PopBack();
+  ASSERT_TRUE(closure.has_value());
+  (*closure)->Run();
+  ASSERT_TRUE(ran);
+  ASSERT_TRUE(queue.Empty());
+}
+
+TEST(WorkQueueTest, PopFrontIsFIFO) {
+  WorkQueue queue;
+  int flag = 0;
+  queue.Add([&flag] { flag |= 1; });
+  queue.Add([&flag] { flag |= 2; });
+  ASSERT_EQ(queue.Size(), 2);
+  (*queue.PopFront())->Run();
+  EXPECT_TRUE(flag & 1);
+  EXPECT_FALSE(flag & 2);
+  (*queue.PopFront())->Run();
+  EXPECT_TRUE(flag & 1);
+  EXPECT_TRUE(flag & 2);
+  ASSERT_TRUE(queue.Empty());
+}
+
+TEST(WorkQueueTest, PopBackIsLIFO) {
+  WorkQueue queue;
+  int flag = 0;
+  queue.Add([&flag] { flag |= 1; });
+  queue.Add([&flag] { flag |= 2; });
+  ASSERT_EQ(queue.Size(), 2);
+  (*queue.PopBack())->Run();
+  EXPECT_FALSE(flag & 1);
+  EXPECT_TRUE(flag & 2);
+  (*queue.PopBack())->Run();
+  EXPECT_TRUE(flag & 1);
+  EXPECT_TRUE(flag & 2);
+  ASSERT_TRUE(queue.Empty());
+}
+
+TEST(WorkQueueTest, OldestEnqueuedTimestampIsSane) {
+  grpc_core::ExecCtx exec_ctx;
+  WorkQueue queue;
+  ASSERT_EQ(queue.OldestEnqueuedTimestamp(), grpc_core::Timestamp::InfPast());
+  queue.Add([] {});
+  ASSERT_LE(queue.OldestEnqueuedTimestamp(), exec_ctx.Now());
+  auto popped = queue.PopFront();
+  ASSERT_EQ(queue.OldestEnqueuedTimestamp(), grpc_core::Timestamp::InfPast());
+  // prevent leaks by executing or deleting the closure
+  delete *popped;
+}
+
+TEST(WorkQueueTest, OldestEnqueuedTimestampOrderingIsCorrect) {
+  WorkQueue queue;
+  grpc_core::ExecCtx exec_ctx;
+  AnyInvocableClosure closure([] {});
+  queue.Add(&closure);
+  absl::SleepFor(absl::Milliseconds(2));
+  queue.Add(&closure);
+  absl::SleepFor(absl::Milliseconds(2));
+  queue.Add(&closure);
+  absl::SleepFor(absl::Milliseconds(2));
+  auto oldest_ts = queue.OldestEnqueuedTimestamp();
+  ASSERT_LE(oldest_ts, exec_ctx.Now());
+  absl::optional<EventEngine::Closure*> popped;
+  // pop the oldest, and ensure the next oldest is younger
+  popped = queue.PopFront();
+  ASSERT_TRUE(popped.has_value());
+  auto second_oldest_ts = queue.OldestEnqueuedTimestamp();
+  ASSERT_GT(second_oldest_ts, oldest_ts);
+  // pop the oldest, and ensure the last one is youngest
+  popped = queue.PopFront();
+  ASSERT_TRUE(popped.has_value());
+  auto youngest_ts = queue.OldestEnqueuedTimestamp();
+  ASSERT_GT(youngest_ts, second_oldest_ts);
+  ASSERT_GT(youngest_ts, oldest_ts);
+}
+
+TEST(WorkQueueTest, ThreadedStress) {
+  WorkQueue queue;
+  constexpr int thd_count = 33;
+  constexpr int element_count_per_thd = 3333;
+  std::vector<std::thread> threads;
+  threads.reserve(thd_count);
+  AnyInvocableClosure closure([] {});
```

Comment:
lets have this closure increment a counter or some such thing, so we can assert that all thd_count*element_count_per_thd closures are scheduled